executor/aggregate.go

```diff
@@ -45,39 +96,484 @@ type HashAggExec struct {
+			finalResultHolderCh: make(chan *chunk.Chunk, 1),
+			rowBuffer:           make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:          chunk.MutRowFromTypes(e.retTypes()),
+		}
+		e.finalWorkers[i].finalResultHolderCh <- e.newChunk()
+	}
+}
+
+func (e *HashAggExec) newFinalAggFuncs() (newAggFuncs []aggregation.Aggregation) {
+	newAggFuncs = make([]aggregation.Aggregation, 0, len(e.AggFuncs))
+	idx := 0
+	for _, af := range e.AggFuncs {
+		var aggFunc aggregation.Aggregation
+		idx, aggFunc = af.GetFinalAggFunc(idx)
+		newAggFuncs = append(newAggFuncs, aggFunc)
+	}
+	return newAggFuncs
+}
+
+// HashAggPartialWorker gets and handles origin data or partial data from inputCh,
+// then shuffle the intermediate results to corresponded final workers.
+func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
+	var (
+		chk         *chunk.Chunk
+		ok          bool
+		sc          = ctx.GetSessionVars().StmtCtx
+		needShuffle = false
+	)
+	defer func() {
+		if needShuffle {
+			w.shuffleIntermData(sc)
+		}
+		waitGroup.Done()
+	}()
+	for {
+		select {
+		case <-w.finishCh:
+			return
+		case chk, ok = <-w.inputCh:
+			if !ok {
+				return
+			}
+		}
+
+		if !w.updateIntermData(ctx, sc, chk, len(w.aggCtxsMaps)) {
+			return
+		}
+		needShuffle = true
+		if chk.NumRows() == 0 {
+			return
+		}
+		w.giveBackCh <- &HashAggInput{
+			chk:        chk,
+			giveBackCh: w.inputCh,
+		}
+	}
+}
+
+func (w *HashAggPartialWorker) updateIntermData(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, finalConcurrency int) (isOk bool) {
+	inputIter := chunk.NewIterator4Chunk(chk)
+	for row := inputIter.Begin(); row != inputIter.End(); row = inputIter.Next() {
+		groupKey, err := w.getGroupKey(sc, row)
+		if err != nil {
+			w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return false
+		}
+		finalWorkerIdx := int(murmur3.Sum32(groupKey)) % finalConcurrency
+		if w.aggCtxsMaps[finalWorkerIdx] == nil {
+			w.aggCtxsMaps[finalWorkerIdx] = make(aggCtxsMapper, 0)
+			w.groupSets[finalWorkerIdx] = mvmap.NewMVMap()
+		}
+		aggCtxMap, groupSet := w.aggCtxsMaps[finalWorkerIdx], w.groupSets[finalWorkerIdx]
+		if len(groupSet.Get(groupKey, w.groupVals[:0])) == 0 {
+			groupSet.Put(groupKey, []byte{})
+		}
+		aggEvalCtxs := w.getContext(sc, groupKey, aggCtxMap)
+		for i, af := range w.aggFuncs {
+			err = af.Update(aggEvalCtxs[i], sc, row)
+			if err != nil {
+				w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// shuffleIntermData shuffles the intermediate data of partial workers to corresponded final workers.
+// We only support parallel execution for single-machine, so process of encode and decode can be skipped.
+func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext) {
+	for i := range w.groupSets {
+		if w.aggCtxsMaps[i] == nil {
+			continue
+		}
+		w.outputChs[i] <- &HashAggIntermData{
+			groupSet:    w.groupSets[i],
```

Comment:
`groupSet` is only used to store all the group key belonging to a final worker? If so, I think `[][]byte` is a better choice, which can help us save the overhead of map access.